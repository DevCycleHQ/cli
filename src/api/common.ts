export const BASE_URL = 'https://api.devcycle.com/'
export const AUTH_URL = 'https://auth.devcycle.com/'

export const buildHeaders = (token: string) => ({
    'Content-Type': 'application/json',
    Authorization: token,
})
