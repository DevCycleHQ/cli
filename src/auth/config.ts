import { Type } from 'class-transformer'
import {
    IsOptional,
    IsString,
    ValidateNested
} from 'class-validator'
import fs from 'fs'
import path from 'path'
import jsYaml from 'js-yaml'

export class ClientCredentialsAuthConfig {
    @IsString()
    client_id: string

    @IsString()
    client_secret: string
}

export class SSOAuthConfig {
    @IsString()
    accessToken: string
}

export class AuthConfig {
    @IsOptional()
    @ValidateNested()
    @Type(() => ClientCredentialsAuthConfig)
    clientCredentials?: ClientCredentialsAuthConfig

    @IsOptional()
    @ValidateNested()
    @Type(() => SSOAuthConfig)
    sso?: SSOAuthConfig
}

export async function storeAccessToken(accessToken:string, authPath:string):Promise<void> {
    const configDir = path.dirname(authPath)
    if (!fs.existsSync(configDir)) {
        fs.mkdirSync(configDir, { recursive: true })
    }
    const config = new AuthConfig()
    config.sso = { accessToken }
    fs.writeFileSync(authPath, jsYaml.dump(config))
}