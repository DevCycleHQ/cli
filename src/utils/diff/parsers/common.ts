import parse from 'parse-diff'
import { ParseOptions, VariableMatch } from './types'

export abstract class BaseParser {
    abstract identity: string
    clientNames: string[]
    abstract variableMethodPattern: RegExp
    abstract variableNameCapturePattern: RegExp

    constructor(protected options: ParseOptions) {
        this.clientNames = options.clientNames ?? ['dvcClient']
    }

    buildRegexPattern() {
        return new RegExp(
            new RegExp(`(?:${this.clientNames.join('|')})`).source
            + this.variableMethodPattern.source
            + this.variableNameCapturePattern.source
        )
    }

    formatMatch(name: string, file: parse.File, change: parse.Change): VariableMatch {
        return {
            name,
            fileName: file.to ?? '',
            line: change.type === 'normal' ? change.ln1 : change.ln,
            mode: change.type === 'add' ? 'add' : 'remove'
        }
    }

    abstract match(content: string, options: ParseOptions): string | null

    parse(file: parse.File, options: ParseOptions = {}): VariableMatch[] {
        const results: VariableMatch[] = []
        for (const chunk of file.chunks) {
            for (const change of chunk.changes) {
                const match = this.match(change.content, options)
                if (match) {
                    results.push(this.formatMatch(match, file, change))
                }
            }
        }
        return results
    }
}
