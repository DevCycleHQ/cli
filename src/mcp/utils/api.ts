import { DevCycleAuth } from './auth'
import { fetchFeatures, createFeature } from '../../api/features'
import { fetchVariables } from '../../api/variables'
import { fetchEnvironmentByKey } from '../../api/environments'
import { enableTargeting, disableTargeting } from '../../api/targeting'

function getErrorMessage(error: unknown): string {
    if (error instanceof Error) {
        return error.message
    }
    return String(error)
}

function ensureError(error: unknown): Error {
    if (error instanceof Error) {
        return error
    }
    return new Error(String(error))
}

export class DevCycleApiClient {
    constructor(private auth: DevCycleAuth) {}

    async listFeatures(args: {
        search?: string
        page?: number
        per_page?: number
    }) {
        console.error('MCP listFeatures params:', JSON.stringify(args, null, 2))

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const authToken = this.auth.getAuthToken()
            const projectKey = this.auth.getProjectKey()

            const query = {
                search: args.search,
                page: args.page,
                perPage: args.per_page,
            }

            const result = await fetchFeatures(authToken, projectKey, query)
            console.error(
                'MCP listFeatures result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP listFeatures error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            console.error('Error during listFeatures:', getErrorMessage(error))
            throw ensureError(error)
        }
    }

    async listVariables(args: {
        search?: string
        page?: number
        per_page?: number
    }) {
        console.error(
            'MCP listVariables params:',
            JSON.stringify(args, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const query = {
                search: args.search,
                page: args.page,
                perPage: args.per_page,
            }

            const result = await fetchVariables(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                query,
            )
            console.error(
                'MCP listVariables result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP listVariables error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }

    async getSdkKeys(
        environmentKey: string,
        keyType?: 'mobile' | 'server' | 'client',
    ) {
        console.error(
            'MCP getSdkKeys params:',
            JSON.stringify({ environmentKey, keyType }, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const environment = await fetchEnvironmentByKey(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                environmentKey,
            )

            const sdkKeys = environment.sdkKeys

            let result
            if (keyType) {
                result = {
                    [keyType]: sdkKeys[keyType],
                }
            } else {
                result = {
                    mobile: sdkKeys.mobile,
                    server: sdkKeys.server,
                    client: sdkKeys.client,
                }
            }

            console.error(
                'MCP getSdkKeys result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP getSdkKeys error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }

    async enableTargeting(featureKey: string, environmentKey: string) {
        console.error(
            'MCP enableTargeting params:',
            JSON.stringify({ featureKey, environmentKey }, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const result = await enableTargeting(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                featureKey,
                environmentKey,
            )
            console.error(
                'MCP enableTargeting result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP enableTargeting error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }

    async disableTargeting(featureKey: string, environmentKey: string) {
        console.error(
            'MCP disableTargeting params:',
            JSON.stringify({ featureKey, environmentKey }, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const result = await disableTargeting(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                featureKey,
                environmentKey,
            )
            console.error(
                'MCP disableTargeting result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP disableTargeting error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }

    async createFeature(args: {
        key?: string
        name?: string
        description?: string
        type?: 'release' | 'experiment' | 'permission' | 'ops'
        interactive?: boolean
    }) {
        console.error(
            'MCP createFeature params:',
            JSON.stringify(args, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            if (args.interactive) {
                // For interactive mode, we would need to implement prompts
                // For now, return an error asking for explicit parameters
                throw new Error(
                    'Interactive mode not yet supported in MCP. Please provide explicit parameters: key, name, description, type',
                )
            }

            if (!args.key || !args.name) {
                throw new Error(
                    'Feature key and name are required when not using interactive mode',
                )
            }

            const featureData = {
                key: args.key,
                name: args.name,
                description: args.description || '',
                type: args.type || 'release',
            }

            const result = await createFeature(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                featureData,
            )
            console.error(
                'MCP createFeature result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP createFeature error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }
}
