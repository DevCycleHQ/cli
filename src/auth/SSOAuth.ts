import crypto, { createHash } from 'crypto'
import http, { IncomingMessage, ServerResponse } from 'http'
import url, { URL } from 'url'

import open from 'open'
import axios from 'axios'
import { Organization } from '../api/organizations'
import Writer from '../ui/writer'
import { toggleBotSadSvg, toggleBotSvg } from '../ui/togglebot'
import { storeAccessToken } from './config'
import { exit } from 'process'

export const CLI_CLIENT_ID = 'Ev9J0DGxR3KhrKaZwY6jlccmjl7JGKEX'

type OauthResponse = {
    data: {
        access_token: string
        refresh_token: string
    }
}
let SUPPORTED_PORTS = [8080, 2194, 2195, 2196]
let PORT = Number.parseInt(process.env.PORT || '2194')

type OauthParams = {
    grant_type: string
    client_id: string
    code_verifier: string
    code: string
    redirect_uri: string
    scope: string
}

type TokenResponse = {
    accessToken: string
    refreshToken: string
    personalAccessToken?: string
}

export default class SSOAuth {
    private organization: Organization | null
    private server: http.Server
    private done: boolean
    private codeVerifier: string
    private tokens: TokenResponse | undefined
    private writer: Writer
    private authPath: string

    constructor(writer: Writer, authPath: string) {
        this.writer = writer
        this.authPath = authPath
    }

    public async getAccessToken(): Promise<Required<TokenResponse>>
    public async getAccessToken(
        organization: Organization,
    ): Promise<Omit<TokenResponse, 'personalAccessToken'>>
    public async getAccessToken(
        organization?: Organization,
    ): Promise<TokenResponse> {
        if (organization) this.organization = organization
        this.startLocalServer()

        const timeout = setTimeout(() => {
            if (!this.done || !this.tokens) {
                this.writer.showError(
                    'Timed out waiting for authentication. Please try again',
                )
                exit(1)
            }
        }, 120000)

        await this.waitForServerClosed()
        const tokens = await this.waitForToken()
        clearTimeout(timeout)
        return tokens
    }

    private async waitForServerClosed(): Promise<void> {
        if (this.done) {
            return
        } else {
            await new Promise((resolve) => setTimeout(resolve, 100))
            return this.waitForServerClosed()
        }
    }

    private async waitForToken(): Promise<TokenResponse> {
        if (this.tokens) {
            return this.tokens
        } else {
            await new Promise((resolve) => setTimeout(resolve, 100))
            return this.waitForToken()
        }
    }

    private startLocalServer() {
        const host = 'localhost'

        if (!SUPPORTED_PORTS.includes(PORT)) {
            throw new Error(
                `Invalid PORT. Only ${SUPPORTED_PORTS.join(', ')} are supported`,
            )
        }

        this.codeVerifier = this.createRandomString()
        const authorizeUrl = this.buildAuthorizeUrl()

        this.server = http.createServer(this.handleAuthRedirect.bind(this))
        this.server.on('error', (e) => {
            console.error(`Local server error ${e}`)
        })
        // prevents keep-alive connections from keeping the server running after close()
        this.server.on('connection', (socket) => {
            socket.unref()
        })

        this.server.on('error', (err: Error & { code: string }) => {
            if (err.code === 'EADDRINUSE') {
                this.writer.showError(
                    `Port ${PORT} already in use. Unable to log in.`,
                )
                SUPPORTED_PORTS = SUPPORTED_PORTS.filter((p) => p !== PORT)
                if (!SUPPORTED_PORTS.length) {
                    this.writer.showError(
                        'No available ports to start local server',
                    )
                    exit(1)
                }
                PORT = SUPPORTED_PORTS[0]
                setTimeout(() => {
                    this.server.close()
                    this.server.listen(PORT, host, () => {
                        this.writer.statusMessage(
                            'Opening browser for authentication...',
                        )
                        open(this.buildAuthorizeUrl())
                    })
                }, 1000)
            } else {
                this.writer.showError(`Error: ${err.message}`)
                exit(1)
            }
        })

        this.server.listen(PORT, host, () => {
            this.writer.statusMessage('Opening browser for authentication...')
            open(authorizeUrl)
        })
        this.server.on('close', this.handleServerClosed.bind(this))
    }

    private handleServerClosed() {
        this.done = true
    }

    private handleAuthRedirect(req: IncomingMessage, res: ServerResponse) {
        const requestUrl = req.url || ''
        const parsed = url.parse(requestUrl, true)
        if (parsed.pathname === '/callback') {
            if (parsed.query.error) {
                res.write(this.resultHtml('Authorization Error', false))
                res.end()
            } else if (parsed.query.code) {
                const code = parsed.query.code.toString()
                this.retrieveAccessToken(code)
                res.write(this.resultHtml('Authorization Successful!', true))
                res.end()
            } else {
                res.write(this.resultHtml('Unrecognized Response', false))
                res.end()
            }
        } else {
            res.writeHead(404)
            res.end()
        }
    }

    private async retrieveAccessToken(code: string) {
        const authHost = 'auth.devcycle.com'
        const host = 'localhost'
        const data: OauthParams = {
            grant_type: 'authorization_code',
            client_id: CLI_CLIENT_ID,
            code_verifier: this.codeVerifier,
            code,
            redirect_uri: `http://${host}:${PORT}/callback`,
            scope: 'offline_access',
        }

        const authUrl = `https://${authHost}/oauth/token`
        const response = await axios
            .post<OauthParams, OauthResponse>(authUrl, data)
            .catch((error) => {
                console.error(error)
            })

        this.server.close()

        if (response?.data) {
            const { access_token, refresh_token } = response.data
            this.tokens = {
                accessToken: access_token,
                refreshToken: refresh_token,
                personalAccessToken: this.organization
                    ? undefined
                    : access_token,
            }
            storeAccessToken(this.tokens, this.authPath)
        }
        if (this.organization) {
            this.writer.successMessage(
                `Access token retrieved for "${this.organization.display_name}" organization`,
            )
        } else {
            this.writer.successMessage('Personal access token retrieved')
        }
    }

    public buildAuthorizeUrl(): string {
        const authHost = 'auth.devcycle.com'
        const host = 'localhost'

        const state = this.createRandomString()
        const code_challenge = createHash('sha256')
            .update(this.codeVerifier)
            .digest('base64url')

        const url = new URL(`https://${authHost}/authorize`)
        url.searchParams.append('response_type', 'code')
        url.searchParams.append('client_id', CLI_CLIENT_ID)
        url.searchParams.append(
            'redirect_uri',
            `http://${host}:${PORT}/callback`,
        )
        url.searchParams.append('state', state)
        url.searchParams.append('code_challenge', code_challenge)
        url.searchParams.append('code_challenge_method', 'S256')
        url.searchParams.append('audience', 'https://api.devcycle.com/')
        url.searchParams.append('scope', 'offline_access')
        if (this.organization) {
            url.searchParams.append('organization', this.organization.id)
        }

        return url.toString()
    }

    private createRandomString() {
        const buffer = crypto.randomBytes(43)
        return buffer.toString('base64url')
    }

    private resultHtml(resultMessage: string, success: boolean) {
        const fontUrl =
            'https://fonts.google.com/share?selection.family=Inter:wght@400;800'
        const backgroundUrl =
            'https://uploads-ssl.webflow.com/614e240a0e0b0fa195b146ed/64b815f3a776eee98d5375a7_backgroundCLI.png'
        return `
<html>
    <head>
        <link href='${fontUrl}' rel='stylesheet' type='text/css'>
        <style>
            #background {
                background-image: url("${backgroundUrl}");
                background-size: cover;
                background-repeat: no-repeat;
                background-position: center;
                display: flex;
                justify-content: center;
                align-items: center;
            }
            h1 {
                font-family: 'Inter', sans-serif;
                font-size: 40px;
                font-weight: 800;
                color: #FFFFFF;
            }
            p {
                font-family: 'Inter', sans-serif;
                font-size: 16px;
                font-weight: 400;
                color: #FFFFFF;
            }
            .container {
                height: fit-content;
                padding-bottom: 2em;
            }
            button {
                background-color: #1D4ED8;
                color: white;
                font-family: 'Inter', sans-serif;
                width: 200px;
                height: 38px;
                border-radius: 6px;
                border-style: none;
                margin-top: 1em;
            }
            button:hover {
                background-color:#335fdb;
                transition: 0.2s;
            }
        </style>
    </head>
    <body align='center' id="background">
        <div class="container">
            ${success ? toggleBotSvg : toggleBotSadSvg}
            <h1>
                ${resultMessage}
            </h1>
            <p>You may now close this page.</p>
        </div>
    </body>
</html>
`
    }
}
