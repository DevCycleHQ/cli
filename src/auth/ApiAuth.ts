import jsYaml from 'js-yaml'
import fs from 'fs'
import axios from 'axios'
import { plainToClass } from 'class-transformer'
import { validateSync } from 'class-validator'
import { AuthConfig, storeAccessToken } from './config'
import { reportValidationErrors } from '../utils/reportValidationErrors'
import { AUTH_URL } from '../api/common'
import { TokenCache } from './TokenCache'
import { shouldRefreshToken } from './utils'
import { CLI_CLIENT_ID } from './SSOAuth'

type SupportedFlags = {
    'client-id'?: string
    'client-secret'?: string
}

export class ApiAuth {
    private authPath: string
    private tokenCache: TokenCache

    constructor(authPath: string, cacheDir: string) {
        this.authPath = authPath
        this.tokenCache = new TokenCache(cacheDir)
    }

    public async getToken(flags: SupportedFlags): Promise<string> {
        const clientId = flags['client-id']
            || process.env.DEVCYCLE_CLIENT_ID
            || process.env.DVC_CLIENT_ID
        const clientSecret = flags['client-secret']
            || process.env.DEVCYCLE_CLIENT_SECRET
            || process.env.DVC_CLIENT_SECRET

        if (clientId && clientSecret) {
            return this.fetchClientToken(clientId, clientSecret)
        }

        if (this.authPath && fs.existsSync(this.authPath)) {
            return this.getTokenFromAuthFile()
        }

        return ''
    }

    private async getTokenFromAuthFile(): Promise<string> {
        const rawConfig = jsYaml.load(fs.readFileSync(this.authPath, 'utf8'))
        const config = plainToClass(AuthConfig, rawConfig)
        const errors = validateSync(config)
        reportValidationErrors(errors)

        if (config.sso) {
            const { accessToken, refreshToken } = config.sso

            if (accessToken && refreshToken && shouldRefreshToken(accessToken)) {
                return this.refreshClientToken(refreshToken)
            }

            return accessToken || ''
        } else if (config.clientCredentials) {
            const { client_id, client_secret } = config.clientCredentials
            return this.fetchClientToken(client_id, client_secret)
        }

        return ''
    }

    private async fetchClientToken(client_id: string, client_secret: string): Promise<string> {
        const cachedToken = this.tokenCache.get(client_id, client_secret)
        if (cachedToken) {
            return cachedToken
        }

        const url = new URL('/oauth/token', AUTH_URL)

        try {
            const response = await axios.post(url.href, {
                grant_type: 'client_credentials',
                client_id,
                client_secret,
                audience: 'https://api.devcycle.com/',
            })

            const accessToken = response.data.access_token
            this.tokenCache.set(client_id, client_secret, accessToken)
            return accessToken
        } catch (e) {
            throw new Error('Failed to authenticate with the DevCycle API. Check your credentials.')
        }
    }

    private async refreshClientToken(refresh_token: string): Promise<string> {
        const url = new URL('/oauth/token', AUTH_URL)

        try {
            const response = await axios.post(url.href, {
                client_id: CLI_CLIENT_ID,
                grant_type: 'refresh_token',
                refresh_token,
                audience: 'https://api.devcycle.com/',
            })

            const accessToken = response.data.access_token
            const refreshToken = response.data.refresh_token
            storeAccessToken({ accessToken, refreshToken }, this.authPath)
            return accessToken
        } catch (e) {
            throw new Error('Failed to refresh DevCycle API token.')
        }
    }
}
