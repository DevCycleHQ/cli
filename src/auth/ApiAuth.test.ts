import * as assert from 'assert'
import { ApiAuth } from './ApiAuth'
import sinon from 'sinon'
import fs from '../utils/fileSystem'
import Writer from '../ui/writer'
import { tokenCacheStub_get } from '../../test/setup'
import test from '@oclif/test'
import { AUTH_URL } from '../api/common'
import { CLI_CLIENT_ID } from './SSOAuth'
import * as config from './config'

const envVars = process.env
const MOCK_AUTH_PATH = 'test-utils/fixtures/auth/mock-auth.yml'

const expiredToken =
    'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6IlhpTzk1Xzllbk53Z1NNSkZRSXZNUiJ9.eyJodHRwczovL2RldmN5Y2xlLmNvbS9vcmdfaWQiOiJvcmdfVTlGOFlNYVRDaFRFbmRXdyIsImlzcyI6Imh0dHBzOi8vYXV0aC5kZXZjeWNsZS5jb20vIiwic3ViIjoiaHNWQm1Scmg1UDlBR2FBSDBlS0dwajMxQ1I1WkxuM3ZAY2xpZW50cyIsImF1ZCI6Imh0dHBzOi8vYXBpLmRldmN5Y2xlLmNvbS8iLCJpYXQiOjE2OTU3NTAxNDUsImV4cCI6MTY5NTgzNjU0NSwiYXpwIjoiaHNWQm1Scmg1UDlBR2FBSDBlS0dwajMxQ1I1WkxuM3YiLCJzY29wZSI6ImNyZWF0ZTpyZXNvdXJjZXMgdXBkYXRlOnJlc291cmNlcyByZWFkOnJlc291cmNlcyBkZWxldGU6cmVzb3VyY2VzIiwiZ3R5IjoiY2xpZW50LWNyZWRlbnRpYWxzIiwicGVybWlzc2lvbnMiOlsiY3JlYXRlOnJlc291cmNlcyIsInVwZGF0ZTpyZXNvdXJjZXMiLCJyZWFkOnJlc291cmNlcyIsImRlbGV0ZTpyZXNvdXJjZXMiXX0.sQn49xJWdcE1xt6r3W8eBrvnUFFP3zOBOJXvjsOOzbfeOPMOeWyR0iqBZ8n96rxtB7wcgR0SB_uk3avNv0zd4X6x-Z4Jv2S0krIPUidTegfO_VLZiSA3uIBiZ--9IZtOLuaQhc16Qq6ezXGibVkFpc6JNLFELVpDQA87pmhF_RoErlH7U_qYuFeYEX_8qksDgeVxYoUOam2O5LksN8BorpzBZ8fikQkHibryqq4MLJvcksDUBwn17H67sGP6wUxWnNbzSksmNqsbKjLSQRvwoQgBL6O6uOW_cGiwnptOcXTz4XC640z2gXRa7FxSHzWKHOGpDRxi6erFSMQIB8_Jow'

afterEach(() => {
    process.env = envVars
    sinon.restore()
})

describe('ApiAuth', () => {
    const mockWriter = new Writer()

    describe('getToken', () => {
        test.nock(AUTH_URL, (api) =>
            api
                .post('/oauth/token', {
                    grant_type: 'client_credentials',
                    client_id: 'flag-client-id',
                    client_secret: 'flag-client-secret',
                    audience: 'https://api.devcycle.com/',
                })
                .reply(200, { access_token: 'mock-token-flag-creds' }),
        ).it(
            'fetches token using client id & secret when passed as flags',
            async () => {
                const auth = new ApiAuth(
                    MOCK_AUTH_PATH,
                    'mock-cache-dir',
                    mockWriter,
                )
                tokenCacheStub_get.returns(null)
                const flags = {
                    'client-id': 'flag-client-id',
                    'client-secret': 'flag-client-secret',
                }
                const response = await auth.getToken(flags)

                assert.equal(response, 'mock-token-flag-creds')
            },
        )

        test.nock(AUTH_URL, (api) =>
            api
                .post('/oauth/token', {
                    grant_type: 'client_credentials',
                    client_id: 'env-client-id',
                    client_secret: 'env-client-secret',
                    audience: 'https://api.devcycle.com/',
                })
                .reply(200, { access_token: 'mock-token-env-creds' }),
        ).it(
            'fetches token using client id & secret when passed as env vars',
            async () => {
                const auth = new ApiAuth(
                    MOCK_AUTH_PATH,
                    'mock-cache-dir',
                    mockWriter,
                )
                process.env.DEVCYCLE_CLIENT_ID = 'env-client-id'
                process.env.DEVCYCLE_CLIENT_SECRET = 'env-client-secret'
                tokenCacheStub_get.returns(null)
                const flags = {}
                const response = await auth.getToken(flags)

                assert.equal(response, 'mock-token-env-creds')
            },
        )

        test.nock(AUTH_URL, (api) =>
            api
                .post('/oauth/token', {
                    grant_type: 'client_credentials',
                    client_id: 'config-client-id',
                    client_secret: 'config-client-secret',
                    audience: 'https://api.devcycle.com/',
                })
                .reply(200, { access_token: 'mock-token-config-creds' }),
        ).it(
            'fetches token using client id & secret from auth config',
            async () => {
                const auth = new ApiAuth(
                    MOCK_AUTH_PATH,
                    'mock-cache-dir',
                    mockWriter,
                )
                tokenCacheStub_get.returns(null)
                sinon.stub(fs, 'existsSync').returns(true)
                sinon.stub(fs, 'readFileSync').returns(
                    JSON.stringify({
                        clientCredentials: {
                            client_id: 'config-client-id',
                            client_secret: 'config-client-secret',
                        },
                    }),
                )
                const flags = {}
                const response = await auth.getToken(flags)

                assert.equal(response, 'mock-token-config-creds')
            },
        )

        it('returns token from cache if available', async () => {
            const auth = new ApiAuth(
                MOCK_AUTH_PATH,
                'mock-cache-dir',
                mockWriter,
            )
            const flags = {
                'client-id': 'mock-client-id',
                'client-secret': 'mock-client-secret',
            }
            tokenCacheStub_get.returns('mock-cached-token')
            const response = await auth.getToken(flags)

            sinon.assert.calledWith(
                tokenCacheStub_get,
                'mock-client-id',
                'mock-client-secret',
            )
            assert.equal(response, 'mock-cached-token')
        })

        it('returns sso token from auth file if available', async () => {
            const auth = new ApiAuth(
                MOCK_AUTH_PATH,
                'mock-cache-dir',
                mockWriter,
            )
            const flags = {}
            sinon.stub(fs, 'existsSync').returns(true)
            sinon.stub(fs, 'readFileSync').returns(
                JSON.stringify({
                    sso: { accessToken: 'mock-config-token' },
                }),
            )

            const response = await auth.getToken(flags)

            assert.equal(response, 'mock-config-token')
        })

        test.nock(AUTH_URL, (api) =>
            api
                .post('/oauth/token', {
                    client_id: CLI_CLIENT_ID,
                    grant_type: 'refresh_token',
                    refresh_token: 'mock-config-refresh-token',
                    audience: 'https://api.devcycle.com/',
                })
                .reply(200, {
                    access_token: 'mock-refreshed-token',
                    refresh_token: 'mock-new-refresh-token',
                }),
        ).it(
            'refreshes sso token from auth file when nearing expiration',
            async () => {
                const auth = new ApiAuth(
                    MOCK_AUTH_PATH,
                    'mock-cache-dir',
                    mockWriter,
                )
                const flags = {}
                sinon.stub(fs, 'existsSync').returns(true)
                sinon.stub(fs, 'readFileSync').returns(
                    JSON.stringify({
                        sso: {
                            accessToken: expiredToken,
                            refreshToken: 'mock-config-refresh-token',
                        },
                    }),
                )
                const storeAccessTokenStub = sinon.stub(
                    config,
                    'storeAccessToken',
                )

                const response = await auth.getToken(flags)

                sinon.assert.calledWith(
                    storeAccessTokenStub,
                    {
                        accessToken: 'mock-refreshed-token',
                        refreshToken: 'mock-new-refresh-token',
                    },
                    MOCK_AUTH_PATH,
                )
                assert.equal(response, 'mock-refreshed-token')
            },
        )

        it('does not refresh sso token if refresh is already in progress', async () => {
            const auth = new ApiAuth(
                MOCK_AUTH_PATH,
                'mock-cache-dir',
                mockWriter,
            )
            auth.refreshInProgress = true
            const flags = {}
            sinon.stub(fs, 'existsSync').returns(true)
            sinon.stub(fs, 'readFileSync').returns(
                JSON.stringify({
                    sso: {
                        accessToken: expiredToken,
                        refreshToken: 'mock-config-refresh-token',
                    },
                }),
            )
            const storeAccessTokenStub = sinon.stub(config, 'storeAccessToken')

            const response = await auth.getToken(flags)

            sinon.assert.notCalled(storeAccessTokenStub)
            assert.equal(response, expiredToken)
        })
    })
})
