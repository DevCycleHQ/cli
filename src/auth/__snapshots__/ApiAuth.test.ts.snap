// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate types correctly generates JS SDK types 1`] = `
"type DVCJSON = { [key: string]: string | boolean | number }

export type DVCVariableTypes = {
    /**
        description: Different ways to say hello
        created by: User 1
        created on: 2021-07-04
    */
    'enum-var': 'Hello' | 'Hey' | 'Hi'
    /**
        created by: User 2
        created on: 2021-07-04
    */
    'regex-var': string
    /**
        created by: User 1
        created on: 2021-07-04
    */
    'string-var': string
    /**
        created by: User 2
        created on: 2021-07-04
    */
    'boolean-var': boolean
    /**
        created by: User 1
        created on: 2021-07-04
    */
    'number-var': number
    /**
        created by: Unknown User
        created on: 2021-07-04
    */
    'json-var': DVCJSON
}

/**
    description: Different ways to say hello
    created by: User 1
    created on: 2021-07-04
*/
export const ENUM_VAR = 'enum-var' as const

/**
    created by: User 2
    created on: 2021-07-04
*/
export const REGEX_VAR = 'regex-var' as const

/**
    created by: User 1
    created on: 2021-07-04
*/
export const STRING_VAR = 'string-var' as const

/**
    created by: User 2
    created on: 2021-07-04
*/
export const BOOLEAN_VAR = 'boolean-var' as const

/**
    created by: User 1
    created on: 2021-07-04
*/
export const NUMBER_VAR = 'number-var' as const

/**
    created by: Unknown User
    created on: 2021-07-04
*/
export const JSON_VAR = 'json-var' as const
"
`;

exports[`generate types correctly generates JS SDK types with inlined comments 1`] = `
"type DVCJSON = { [key: string]: string | boolean | number }

export type DVCVariableTypes = {
    'enum-var': 'Hello' | 'Hey' | 'Hi' // (Different ways to say hello) created by User 1 on 2021-07-04
    'regex-var': string // created by User 2 on 2021-07-04
    'string-var': string // created by User 1 on 2021-07-04
    'boolean-var': boolean // created by User 2 on 2021-07-04
    'number-var': number // created by User 1 on 2021-07-04
    'json-var': DVCJSON // created by Unknown User on 2021-07-04
}

 // (Different ways to say hello) created by User 1 on 2021-07-04
export const ENUM_VAR = 'enum-var' as const

 // created by User 2 on 2021-07-04
export const REGEX_VAR = 'regex-var' as const

 // created by User 1 on 2021-07-04
export const STRING_VAR = 'string-var' as const

 // created by User 2 on 2021-07-04
export const BOOLEAN_VAR = 'boolean-var' as const

 // created by User 1 on 2021-07-04
export const NUMBER_VAR = 'number-var' as const

 // created by Unknown User on 2021-07-04
export const JSON_VAR = 'json-var' as const
"
`;

exports[`generate types correctly generates React SDK types 1`] = `
"import { DVCVariable, DVCVariableValue } from '@devcycle/js-client-sdk'
import {
    useVariable as originalUseVariable,
    useVariableValue as originalUseVariableValue
} from '@devcycle/react-client-sdk'

type DVCJSON = { [key: string]: string | boolean | number }

export type DVCVariableTypes = {
    /**
        description: Different ways to say hello
        created by: User 1
        created on: 2021-07-04
    */
    'enum-var': 'Hello' | 'Hey' | 'Hi'
    /**
        created by: User 2
        created on: 2021-07-04
    */
    'regex-var': string
    /**
        created by: User 1
        created on: 2021-07-04
    */
    'string-var': string
    /**
        created by: User 2
        created on: 2021-07-04
    */
    'boolean-var': boolean
    /**
        created by: User 1
        created on: 2021-07-04
    */
    'number-var': number
    /**
        created by: Unknown User
        created on: 2021-07-04
    */
    'json-var': DVCJSON
}

/**
    description: Different ways to say hello
    created by: User 1
    created on: 2021-07-04
*/
export const ENUM_VAR = 'enum-var' as const

/**
    created by: User 2
    created on: 2021-07-04
*/
export const REGEX_VAR = 'regex-var' as const

/**
    created by: User 1
    created on: 2021-07-04
*/
export const STRING_VAR = 'string-var' as const

/**
    created by: User 2
    created on: 2021-07-04
*/
export const BOOLEAN_VAR = 'boolean-var' as const

/**
    created by: User 1
    created on: 2021-07-04
*/
export const NUMBER_VAR = 'number-var' as const

/**
    created by: Unknown User
    created on: 2021-07-04
*/
export const JSON_VAR = 'json-var' as const

export type UseVariableValue = <
    K extends string & keyof DVCVariableTypes
>(
    key: K,
    defaultValue: DVCVariableTypes[K]
) => DVCVariableTypes[K]

export const useVariableValue: UseVariableValue = originalUseVariableValue

export type UseVariable = <
    K extends string & keyof DVCVariableTypes,
    T extends DVCVariableValue & DVCVariableTypes[K],
>(
    key: K,
    defaultValue: DVCVariableTypes[K]
) => DVCVariable<T>

export const useVariable: UseVariable = originalUseVariable"
`;
